org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -495,25 +496,47 @@ public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
 				if (currentLS != null && currentLS == LanguageServerWrapper.this.languageServer) {
 					currentLS.getWorkspaceService().didChangeWorkspaceFolders(new DidChangeWorkspaceFoldersParams(wsFolderEvent));
 				}
-				ResourcesPlugin.getWorkspace().addResourceChangeListener(workspaceFolderUpdater, IResourceChangeEvent.POST_CHANGE);
+				ResourcesPlugin.getWorkspace().addResourceChangeListener(workspaceFolderUpdater, IResourceChangeEvent.POST_CHANGE | IResourceChangeEvent.PRE_DELETE);
 				return Status.OK_STATUS;
 			}
 		}.schedule();
 	}
 
-	private static final @Nullable WorkspaceFoldersChangeEvent toWorkspaceFolderEvent(IResourceChangeEvent e) {
-		if (e.getType() != IResourceChangeEvent.POST_CHANGE) {
+	private static final @Nullable WorkspaceFoldersChangeEvent toWorkspaceFolderEvent(
+			IResourceChangeEvent e) {
+
+		// If a project delete then the delta is null, but we get the project in the top-level resource
+		WorkspaceFoldersChangeEvent wsFolderEvent = new WorkspaceFoldersChangeEvent();
+		if (e.getType() == IResourceChangeEvent.PRE_DELETE) {
+			final IResource resource = e.getResource();
+			if (resource instanceof IProject) {
+				wsFolderEvent.getRemoved()
+						.add(LSPEclipseUtils.toWorkspaceFolder((IProject)resource));
+				return wsFolderEvent;
+			} else {
+				return null;
+			}
+		} else if (e.getType() != IResourceChangeEvent.POST_CHANGE) {
 			return null;
 		}
-		WorkspaceFoldersChangeEvent wsFolderEvent = new WorkspaceFoldersChangeEvent();
+
+		// Use the visitor implementation to extract the low-level detail from delta
 		try {
 			e.getDelta().accept(delta -> {
 				if (delta.getResource().getType() == IResource.PROJECT) {
-					IProject project = (IProject)delta.getResource();
-					if ((delta.getKind() == IResourceDelta.ADDED || delta.getKind() == IResourceDelta.OPEN) && project.isAccessible()) {
-						wsFolderEvent.getAdded().add(LSPEclipseUtils.toWorkspaceFolder((IProject)delta.getResource()));
-					} else if (delta.getKind() == IResourceDelta.REMOVED || (delta.getKind() == IResourceDelta.OPEN && !project.isAccessible())) {
-						wsFolderEvent.getRemoved().add(LSPEclipseUtils.toWorkspaceFolder((IProject)delta.getResource()));
+					IProject project = (IProject) delta.getResource();
+					final WorkspaceFolder wsFolder = LSPEclipseUtils.toWorkspaceFolder(project);
+					if ((delta.getKind() == IResourceDelta.ADDED || (delta.getKind() == IResourceDelta.CHANGED
+							&& (delta.getFlags() & IResourceDelta.OPEN) == IResourceDelta.OPEN))
+							&& project.isAccessible()
+							&& wsFolder != null && !wsFolder.getUri().isEmpty()) {
+						wsFolderEvent.getAdded().add(wsFolder);
+					} else if ((delta.getKind() == IResourceDelta.REMOVED
+							|| (delta.getKind() == IResourceDelta.CHANGED
+									&& (delta.getFlags() & IResourceDelta.OPEN) == IResourceDelta.OPEN))
+									&& !project.isAccessible()
+									&& wsFolder != null && !wsFolder.getUri().isEmpty()) {
```

Comment:
this condition and the previous one are almost identical, could we refactor to a private method with a boolean flag `isAccessbile`?

Reply:
It would be nice to break up the code a bit, but this class is already rather large, and so is e.g. `LSPEclipseUtils`. What would you say to replacing the lambda `workspaceFolderUpdater` at line 172 with a proper class implementing `IResourceChangeListener`, then all of this logic could be broken up and put there?

Reply:
I have no strong opinion about refactoring to a nested class. Regarding the null check, I think you need it in any case because when a project is deleted the URI will be null. I think this can happen regardless of the de-registration in `stop()`.